text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+            "protocol_max": "<max TLS version>",
+            "ciphers": "<available ciphers>"
+        },
+        ...
+    }
+
+If an instance has no record in ``config/tls.json``, the EdgeDB server options
+``--tls-protocol-min``, ``--tls-protocol-max`` and ``--tls-ciphers`` will be
+omitted and the default value will be used.
+
+``credentials/<instance>.json`` is updated to include the passphrase for the
+private key file, if any. For example::
+
+    {
+        "port": 10732,
+        "user": "edgedb",
+        "password": "....",
+        "database": "edgedb",
+        "tls_private_key_password": "...."
+    }
+
+Question: I don't feel comfortable storing the key passphrase here in clear
+text, especially if this is in production environment and we asked the user
+for passphrase but stored in this file. Perhaps we could use this JSON value
+only for the development process, and introduce a more secure (thus less
+convenient) way to obtain the private key password, like a customizable command
+as in PostgreSQL?
+
+Under the new folder ``certs/``, we'll place all certificates and keys used by
+the EdgeDB instances. The cert files will be named after the instance names.
+For development scenario we'll likely generate those certs for the user, while
+in production we'll "import" the certs specified by the user, by copying and
+renaming the cert and key files to the ``certs/`` folder.
+
+
+Development Scenario
+^^^^^^^^^^^^^^^^^^^^
+
+From the user's perspective, ``edgedb server init`` (without parameters we are
+introducing in this RFC later) will now automatically generate a self-signed
+certificate with a random key passphrase, and set the above config properly.
+This is considered to be the development scenario, so we try not to let the
+user do extra work here - the ``edgedb server init`` command shouldn't ask
+questions about generating the self-signed certificate. Instead, it should warn
+about the self-signed certificate, and provide suggestions for production.
+
+Question: because the certificate is self-signed, certificate verification on
+the client side would fail. Now that we know we are "in development mode"
+(using self-signed certificate), we could leverage the tooling to trust the
+certificate on the client side, by adding a boolean flag in
+``$WELLKNOWN/credentials/<instance>.json``. We could update our client bindings
+and the CLI to honor this flag, load and trust the certificate so that the user
+doesn't have to write extra code in development. Though, this works only for
+local clients - remote clients don't have access to the local file system so
+there has to be a different way for them to trust the dev server. If this
+option is good, I will expand ths paragraph with details. Alternatives:
+
+1. Automatically build a local CA, and issue certificates from that CA. The
+   clients and CLI will only need to trust the CA certificate. This doesn't
+   solve the remote client trouble though.
```

Comment:
Yeah, I think this should be the default way to go for both `edb server` and `edgedb server` workflows.

Reply:
Works for me 👍 And for the implementation, I think both could use a new command like `edb ca` that uses OpenSSL (new dependency) to operate the CA and issue dev certs. (No more rustls?)

Reply:
Rustls has CA functionality. But do you mean `edgedb ca` ? (`edb` is a developement command, I don't think you propose to bundle some CA key in to the distribution itself, right?)

I don't think we need CA. Just put self-signed certificate (public part) into a `~/.config/edgedb` folder somewhere and use that in-client.

Reply:
> But do you mean `edgedb ca` ?

Oh sorry I meant the general Python code, I was confused. You're right, we could use the CLI to operate the CA - not to say I think we need one.